File: #114_Flatten_Binary_Tree_to_Linked_List/Solution.java
/* Use similar idea with Morris traversal */

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
    public void flatten(TreeNode root) {
		TreeNode cur = root;
		while (cur != null) {
			if (cur.left != null) {
				TreeNode last = cur.left;
				while (last.right != null) {
					last = last.right;
				}
				last.right = cur.right;
				cur.right = cur.left;
				cur.left = null;
			}
			cur = cur.right;
		}
    }
}

File: #114_Flatten_Binary_Tree_to_Linked_List/Solution.py
'''
Recursive pre-order Solution
takes advantage of the fact that flattened tree are actually in pre-order
Takes time O(n)
sapce O(log(n))
'''

class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:

	def __init__(self):
		self.cur = None

	# @param {TreeNode} root
	# @return {void} Do not return anything, modify root in-place instead.
	def flatten(self, root):
		if not root:
			return
		self.cur = root
		self.flatten(root.left)
		self.cur.right = root.right
		if root.left:
			root.right = root.left
		root.left = None
		self.flatten(self.cur.right)


sol = Solution()
root = TreeNode(0)
n1 = TreeNode(1)
root.left = n1
sol.flatten(root)